### PR TITLE
Ssp common module - codec type detection

### PR DESCRIPTION
**Pull request by brentlu**
Create an new common module to host functionalities for all ssp machine drivers:
1. codec and amplifier type detection
2. DAI link generation

Machine drivers sof_rt5682, sof_cs42l42, and sof_ssp_amp are tested on ADL Chromebooks. Check headphone playback/capture, speaker playback, DMIC capture, and HDMI playback function.

Brya
[    5.941973] sof_rt5682 adl_mx98360_rt5682: codec RT5682 found
[    5.941986] sof_rt5682 adl_mx98360_rt5682: amp MAX98360A found
[    5.941988] sof_rt5682 adl_mx98360_rt5682: create dai links, link_order 0x7654321
[    5.941989] sof_rt5682 adl_mx98360_rt5682: link 0: codec RT5682, ssp 0
[    5.941990] sof_rt5682 adl_mx98360_rt5682: link 1: dmic01
[    5.941991] sof_rt5682 adl_mx98360_rt5682: link 2: dmic16k
[    5.941991] sof_rt5682 adl_mx98360_rt5682: link 3: intel hdmi, hdmi id 1
[    5.941992] sof_rt5682 adl_mx98360_rt5682: link 4: intel hdmi, hdmi id 2
[    5.941993] sof_rt5682 adl_mx98360_rt5682: link 5: intel hdmi, hdmi id 3
[    5.941994] sof_rt5682 adl_mx98360_rt5682: link 6: intel hdmi, hdmi id 4
[    5.941995] sof_rt5682 adl_mx98360_rt5682: link 7: amp MAX98360A, ssp 1
[    5.941996] sof_rt5682 adl_mx98360_rt5682: link 8: bt, ssp 2

Vell
[    6.475233] sof_ssp_amp adl_cs35l41: amp CS35L41 found
[    6.475240] sof_ssp_amp adl_cs35l41: create dai links, link_order 0x643257
[    6.475243] sof_ssp_amp adl_cs35l41: link 0: amp CS35L41, ssp 1
[    6.475265] sof_ssp_amp adl_cs35l41: link 1: dmic01
[    6.475266] sof_ssp_amp adl_cs35l41: link 2: dmic16k
[    6.475268] sof_ssp_amp adl_cs35l41: link 3: intel hdmi, hdmi id 1
[    6.475270] sof_ssp_amp adl_cs35l41: link 4: intel hdmi, hdmi id 2
[    6.475272] sof_ssp_amp adl_cs35l41: link 5: intel hdmi, hdmi id 3
[    6.475274] sof_ssp_amp adl_cs35l41: link 6: intel hdmi, hdmi id 4
[    6.475276] sof_ssp_amp adl_cs35l41: link 7: bt, ssp 2

Crota
[    7.727384] sof_cs42l42 adl_mx98360a_cs4242: codec CS42L42 found
[    7.727414] sof_cs42l42 adl_mx98360a_cs4242: amp MAX98360A found
[    7.727416] sof_cs42l42 adl_mx98360a_cs4242: create dai links, link_order 0x7654321
[    7.727418] sof_cs42l42 adl_mx98360a_cs4242: link 0: codec CS42L42, ssp 0
[    7.727420] sof_cs42l42 adl_mx98360a_cs4242: link 1: dmic01
[    7.727421] sof_cs42l42 adl_mx98360a_cs4242: link 2: dmic16k
[    7.727422] sof_cs42l42 adl_mx98360a_cs4242: link 3: intel hdmi, hdmi id 1
[    7.727424] sof_cs42l42 adl_mx98360a_cs4242: link 4: intel hdmi, hdmi id 2
[    7.727426] sof_cs42l42 adl_mx98360a_cs4242: link 5: intel hdmi, hdmi id 3
[    7.727428] sof_cs42l42 adl_mx98360a_cs4242: link 6: intel hdmi, hdmi id 4
[    7.727430] sof_cs42l42 adl_mx98360a_cs4242: link 7: amp MAX98360A, ssp 1
[    7.727434] sof_cs42l42 adl_mx98360a_cs4242: link 8: bt, ssp 2

### CODE REVIEW

File: sound/soc/intel/boards/sof_realtek_common.h [hunk 1]
```diff
@@ -12,35 +12,55 @@
...
+#define RT1308_DEV0_NAME	"i2c-" RT1308_ACPI_HID ":00"
 void sof_rt1308_dai_link(struct snd_soc_dai_link *link);
 
+/*
+ * Realtek ALC1019
+ */
+#define RT1019P_ACPI_HID	"RTL1019"
 #define RT1019P_CODEC_DAI	"HiFi"
-#define RT1019P_DEV0_NAME	"RTL1019:00"
+#define RT1019P_DEV0_NAME	RT1019P_ACPI_HID ":00"
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_realtek_common.h:60]: this sort of small code changes should be in a separate patch marked with 'no functional change' to make the reviews easier.

File: sound/soc/intel/boards/sof_ssp_common.c [hunk 2]
```diff
@@ -0,0 +1,101 @@
+// SPDX-License-Identifier: GPL-2.0-only
+//
+// Copyright(c) 2023 Intel Corporation. All rights reserved.
+
+#include <linux/device.h>
+#include <sound/soc-acpi.h>
+#include "sof_cirrus_common.h"
+#include "sof_maxim_common.h"
+#include "sof_realtek_common.h"
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_ssp_common.c]: do we really need to include all these vendor-specific headers? That sounds like a partitioning of code that's not quite right?

File: sound/soc/intel/boards/Kconfig [hunk 3]
```diff
@@ -31,15 +31,29 @@ config SND_SOC_INTEL_HDA_DSP_COMMON
 
 config SND_SOC_INTEL_SOF_MAXIM_COMMON
 	tristate
+	select SND_SOC_MAX98357A
+	select SND_SOC_MAX98373_I2C
+	select SND_SOC_MAX98390
```
Comment (plbossart) [inline sound/soc/intel/boards/Kconfig]: this is very odd and defeats the purpose of isolating the 'common' parts. I am not following what you are trying to do and why this is necessary. We should only select the minimal configurations.

File: sound/soc/intel/boards/Kconfig [hunk 4]
```diff
@@ -31,15 +31,29 @@ config SND_SOC_INTEL_HDA_DSP_COMMON
...
+	select SND_SOC_CS35L41_I2C
 
 config SND_SOC_INTEL_SOF_SSP_COMMON
 	tristate
+	select SND_SOC_DMIC
+	select SND_SOC_HDAC_HDMI
+	select SND_SOC_INTEL_HDA_DSP_COMMON
+	select SND_SOC_INTEL_SOF_CIRRUS_COMMON
+	select SND_SOC_INTEL_SOF_MAXIM_COMMON
+	select SND_SOC_INTEL_SOF_REALTEK_COMMON
```
Comment (plbossart) [inline sound/soc/intel/boards/Kconfig]: yeah, that seems wrong. SSP should never have to select HDAC_HDMI or DMIC. I am not in agreement here.

File: sound/soc/intel/boards/sof_ssp_amp.c [hunk 5]
```diff
@@ -499,6 +238,4 @@ MODULE_DESCRIPTION("ASoC Intel(R) SOF Amplifier Machine driver");
 MODULE_AUTHOR("Balamurugan C <balamurugan.c@intel.com>");
 MODULE_AUTHOR("Brent Lu <brent.lu@intel.com>");
 MODULE_LICENSE("GPL");
-MODULE_IMPORT_NS(SND_SOC_INTEL_HDA_DSP_COMMON);
-MODULE_IMPORT_NS(SND_SOC_INTEL_SOF_REALTEK_COMMON);
-MODULE_IMPORT_NS(SND_SOC_INTEL_SOF_CIRRUS_COMMON);
+MODULE_IMPORT_NS(SND_SOC_INTEL_SOF_SSP_COMMON);
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_ssp_amp.c:521]: common comment for all machine changes: this PR is a bit too invasive. 

The notion of detecting codecs with ACPI is fine, it's the right thing to do. Can we start with this first? 

I don't see the point of trying to change DMIC and HDMI support at the same time, and adding common parts to model different dailink orders. This is starting to bring limited returns compared to lots of code changes. 

Can we start with the codec detection first, and limit the PR to this? It'll be a lot easier to review and merge.

**Review by plbossart — CHANGES_REQUESTED** [review]
The idea of removing codec quirks is good, but I think you went too far in the refactoring of codec @brentlu

File: sound/soc/intel/boards/sof_ssp_common.c (hunk 2 above)
Comment (brentlu) [reply sound/soc/intel/boards/sof_ssp_common.c]: The ACPI HID of amplifiers are defined in sof_vendor_common.h so those headers are included here.

**Comment by brentlu** [comment]
Upload new patchset to support codec detection only

File: sound/soc/intel/boards/sof_realtek_common.h (hunk 1 above)
Comment (plbossart) [reply sound/soc/intel/boards/sof_realtek_common.h:60]: comment is still valid, please move this to a 'no functional change' separate patch

File: sound/soc/intel/boards/sof_ssp_common.c (hunk 2 above)
Comment (plbossart) [reply sound/soc/intel/boards/sof_ssp_common.c]: comment is still valid, I think you just need to include sof_vendor_common.h ?

**Review by plbossart — COMMENTED** [review]
I only have minor comments below @brentlu but that PR is moving in the right direction. Removing these quirks is a good thing (tm), thanks for suggesting it.

@bardliao @RanderWang please review if you have spare cycles

File: sound/soc/intel/boards/sof_realtek_common.h (hunk 1 above)
Comment (brentlu) [reply sound/soc/intel/boards/sof_realtek_common.h:60]: done

File: sound/soc/intel/boards/sof_rt5682.c [hunk 6]
```diff
@@ -120,30 +113,13 @@ static const struct dmi_system_id sof_rt5682_quirk_table[] = {
...
-		.callback = sof_rt5682_quirk_cb,
-		.matches = {
-			DMI_MATCH(DMI_SYS_VENDOR, "HP"),
-			DMI_MATCH(DMI_PRODUCT_NAME, "Dooly"),
-		},
-		.driver_data = (void *)(SOF_RT5682_MCLK_EN |
-					SOF_RT5682_SSP_CODEC(0) |
-					SOF_SPEAKER_AMP_PRESENT |
-					SOF_RT1015_SPEAKER_AMP_PRESENT |
-					SOF_RT5682_SSP_AMP(1)),
```
Comment (bardliao) [inline sound/soc/intel/boards/sof_rt5682.c:137]: I know `SOF_RT5682_MCLK_EN` and `SOF_RT5682_SSP_CODEC(0)` are the default value, but can we remove `SOF_RT5682_SSP_AMP(1)`? How about just remove `SOF_SPEAKER_AMP_PRESENT` and `SOF_RT1015_SPEAKER_AMP_PRESENT`?

File: sound/soc/intel/boards/sof_rt5682.c [hunk 7]
```diff
@@ -1038,15 +1050,51 @@ static int sof_audio_probe(struct platform_device *pdev)
...
-	if (sof_rt5682_quirk & SOF_RT1015_SPEAKER_AMP_PRESENT)
-		sof_rt1015_codec_conf(&sof_audio_card_rt5682);
+	/* update codec_conf */
+	if (ctx->amp_type != CODEC_NONE) {
+		switch (ctx->amp_type) {
+		case CODEC_MAX98357A:
+			/* no codec conf required */
+			break;
+		case CODEC_MAX98360A:
+			/* no codec conf required */
```
Comment (bardliao) [inline sound/soc/intel/boards/sof_rt5682.c]: nitpick: We can group all those codecs together as
```
case CODEC_MAX98357A:
case CODEC_MAX98360A:
case CODEC_RT1019P:
case CODEC_RT5650:
	/* no codec conf required */
	break;
```

**Review by bardliao — COMMENTED** [review]
Thanks @brentlu. A question about removing SOF_RT5682_SSP_AMP(1). Is it intentional?

File: sound/soc/intel/boards/sof_rt5682.c (hunk 6 above)
Comment (brentlu) [reply sound/soc/intel/boards/sof_rt5682.c:137]: CML chromebooks use "Google_Hatch" quirk which amplifier is MAX98357A so we added a quirk for Dooly to use RT1015 amp. Since we could detect the amp type now, this quirk could be removed.

Comment (bardliao) [reply sound/soc/intel/boards/sof_rt5682.c:137]: > CML chromebooks use "Google_Hatch" quirk which amplifier is MAX98357A so we added a quirk for Dooly to use RT1015 amp. Since we could detect the amp type now, this quirk could be removed.

Thanks for the explanation. So, Dooly will use "Google_Hatch" quirk.

**Review by bardliao — DISMISSED** [review]

File: sound/soc/intel/boards/sof_ssp_common.c (hunk 2 above)
Comment (brentlu) [reply sound/soc/intel/boards/sof_ssp_common.c]: I mean the sof_cirrus_common.h, sof_maxim_common.h, and sof_realtek_common.h. Unless we define the HID definition in sof_ssp_common.h and leave those file untouched.

File: sound/soc/intel/boards/sof_ssp_common.h [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+
+#ifndef __SOF_SSP_COMMON_H
+#define __SOF_SSP_COMMON_H
+
+#include "sof_cirrus_common.h"
+#include "sof_maxim_common.h"
+#include "sof_realtek_common.h"
+
+/* Cirrus Logic */
+#define CS42L42_ACPI_HID	"10134242"
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_ssp_common.h:11]: That's inconsistent, in the first patch you put an ACPI_HID in cirrus_common:
https://github.com/thesofproject/linux/pull/4536/commits/49c6579b2596cdf102d4d837f0e453ace148988e#diff-cc3ecb492eb6b9063ede92817592e3813d707c4fdbb77b6dbe7fd433c2fba844R16

and here it's different.

We really want all the HIDs in a single place.

File: sound/soc/intel/boards/sof_rt5682.c (hunk 7 above)
Comment (plbossart) [reply sound/soc/intel/boards/sof_rt5682.c]: yes, agree with @bardliao let's have a common block here.

**Review by plbossart — CHANGES_REQUESTED** [review]
Almost good @brentlu, I added two comments below. One more round and we should be good to go.

File: sound/soc/intel/boards/sof_ssp_common.h [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+#define RT1015_ACPI_HID		"10EC1015"
+#define RT1015P_ACPI_HID	"RTL1015"
+#define RT1019P_ACPI_HID	"RTL1019"
+#define RT1308_ACPI_HID		"10EC1308"
+#define RT5650_ACPI_HID		"10EC5650"
+#define RT5682_ACPI_HID		"10EC5682"
+#define RT5682S_ACPI_HID	"RTL5682"
+
+enum sof_ssp_codec {
+	CODEC_NONE = 0,
```
Comment (ranj063) [inline sound/soc/intel/boards/sof_ssp_common.h]: is the = 0 really important?

File: sound/soc/intel/boards/sof_nau8825.c [hunk 10]
```diff
@@ -581,24 +579,45 @@ static int sof_audio_probe(struct platform_device *pdev)
...
+			break;
+		case CODEC_NAU8318:
+			/* no codec conf required */
+			break;
+		case CODEC_RT1015P:
+			sof_rt1015p_codec_conf(&sof_audio_card_nau8825);
+			break;
+		case CODEC_RT1019P:
+			/* no codec conf required */
+			break;
```
Comment (plbossart) [inline sound/soc/intel/boards/sof_nau8825.c]: nit-pick: can you group the 'no codec conf required' as done in the earlier patches?

**Review by plbossart — CHANGES_REQUESTED** [review]
one more cosmetic comment from me, thanks @brentlu

File: sound/soc/intel/boards/sof_nau8825.c (hunk 10 above)
Comment (brentlu) [reply sound/soc/intel/boards/sof_nau8825.c]: done

File: sound/soc/intel/boards/sof_rt5682.c (hunk 7 above)
Comment (brentlu) [reply sound/soc/intel/boards/sof_rt5682.c]: done

File: sound/soc/intel/boards/sof_ssp_common.h (hunk 9 above)
Comment (brentlu) [reply sound/soc/intel/boards/sof_ssp_common.h]: According to the bible, "if no enumerators with = appear, then the values of the corresponding constants begin at 0 and increase by 1 as the declaration is read from left to right". So I think you are right, the = 0 here is more than a habit. Will remove it.

**Review by plbossart — DISMISSED** [review]
LGTM, nice cleanup @brentlu

**Review by plbossart — DISMISSED** [review]

**Review by bardliao — DISMISSED** [review]

File: sound/soc/intel/boards/sof_ssp_common.c [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+
+#include <linux/device.h>
+#include <sound/soc-acpi.h>
+#include "sof_ssp_common.h"
+
+/*
+ * Codec probe function
+ */
+#define CODEC_MAP_ENTRY(n, h, t)	\
+	{					\
```
Comment (RanderWang) [inline sound/soc/intel/boards/sof_ssp_common.c]: align the  \

File: sound/soc/intel/boards/sof_nau8825.c [hunk 12]
```diff
@@ -581,24 +579,41 @@ static int sof_audio_probe(struct platform_device *pdev)
...
+	dai_links = sof_card_dai_links_create(&pdev->dev, ctx->amp_type,
+					      ssp_codec, ssp_amp, dmic_be_num,
+					      hdmi_num);
 	if (!dai_links)
 		return -ENOMEM;
 
 	sof_audio_card_nau8825.dai_link = dai_links;
 
+	/* update codec_conf */
+	if (ctx->amp_type != CODEC_NONE) {
```
Comment (RanderWang) [inline sound/soc/intel/boards/sof_nau8825.c]: why not remove this if and use it like
````c
switch (ctx->amp_type) {
                ...
		case CODEC_NONE:
                break;
                default:
                ...
}
````

File: sound/soc/intel/boards/sof_ssp_amp.c [hunk 13]
```diff
@@ -413,15 +420,28 @@ static int sof_ssp_amp_probe(struct platform_device *pdev)
...
+					      ctx->idisp_codec);
 	if (!dai_links)
 		return -ENOMEM;
 
 	sof_ssp_amp_card.dai_link = dai_links;
 
 	/* update codec_conf */
-	if (sof_ssp_amp_quirk & SOF_CS35L41_SPEAKER_AMP_PRESENT) {
-		cs35l41_set_codec_conf(&sof_ssp_amp_card);
+	if (ctx->amp_type != CODEC_NONE) {
```
Comment (RanderWang) [inline sound/soc/intel/boards/sof_ssp_amp.c]: convert this if to case

File: sound/soc/intel/boards/sof_ssp_common.c [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+enum sof_ssp_codec sof_ssp_detect_amp_type(struct device *dev)
+{
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(amps); i++) {
+		if (!acpi_dev_present(amps[i].acpi_hid, NULL, -1))
+			continue;
+
+		dev_dbg(dev, "amp %s found\n", amps[i].name);
+		return amps[i].codec_type;
```
Comment (RanderWang) [inline sound/soc/intel/boards/sof_ssp_common.c:92]: is this equal to 
````c
if (acpi_dev_present(amps[i].acpi_hid, NULL, -1)) {
    dev_dbg(dev, "amp %s found\n", amps[i].name);
    return amps[i].codec_type;
}
````

**Review by RanderWang — COMMENTED** [review]
good work!  it can be better.

File: sound/soc/intel/boards/sof_ssp_common.c (hunk 14 above)
Comment (brentlu) [reply sound/soc/intel/boards/sof_ssp_common.c:92]: Yeah, they are equal. I just use style with less indent.

File: sound/soc/intel/boards/sof_ssp_common.c (hunk 11 above)
Comment (brentlu) [reply sound/soc/intel/boards/sof_ssp_common.c]: fixed

File: sound/soc/intel/boards/sof_nau8825.c (hunk 12 above)
Comment (brentlu) [reply sound/soc/intel/boards/sof_nau8825.c]: fixed

File: sound/soc/intel/boards/sof_ssp_amp.c (hunk 13 above)
Comment (brentlu) [reply sound/soc/intel/boards/sof_ssp_amp.c]: fixed

**Review by RanderWang — APPROVED** [review]
LGTM

**Review by plbossart — APPROVED** [review]

**Review by bardliao — APPROVED** [review]

**Comment by brentlu** [comment]
Could we merge this PR? Thanks.

**Comment by plbossart** [comment]
merging but we'll send this upstream after the merge window closes in two weeks.
